File: typescript/packages/ampersend-sdk/src/x402/wallets/smart-account/wallet.ts
import { type Address, type Hex } from "viem"

import { COSIGNER_VALIDATOR, OWNABLE_VALIDATOR } from "../../../smart-account/constants.ts"
import { acceptedOf, type PaymentAuthorization, type PaymentInstruction } from "../../envelopes.ts"
import type { ServerAuthorizationData } from "../../types.ts"
import { WalletError, type X402Wallet } from "../../wallet.ts"
import { createCoSignedPayment } from "./cosigned.ts"
import { createExactPayment } from "./exact.ts"

export interface SmartAccountConfig {
  smartAccountAddress: Address
  sessionKeyPrivateKey: Hex
  /** OwnableValidator address. Defaults to the standard OwnableValidator. */
  validatorAddress?: Address
  /** CoSignerValidator address. Defaults to the standard CoSignerValidator. */
  coSignerValidatorAddress?: Address
}

/**
 * Smart account wallet using ERC-1271. Supports Safe accounts with the
 * OwnableValidator module and only the `exact` scheme (ERC-3009 / USDC).
 */
export class SmartAccountWallet implements X402Wallet {
  private readonly config: SmartAccountConfig & { validatorAddress: Address; coSignerValidatorAddress: Address }

  constructor(config: SmartAccountConfig) {
    this.config = {
      ...config,
      validatorAddress: config.validatorAddress ?? OWNABLE_VALIDATOR,
      coSignerValidatorAddress: config.coSignerValidatorAddress ?? COSIGNER_VALIDATOR,
    }
  }

  async createPayment(
    instruction: PaymentInstruction,
    serverAuthorization?: ServerAuthorizationData,
  ): Promise<PaymentAuthorization> {
    const accepted = acceptedOf(instruction)
    if (accepted.scheme !== "exact") {
      throw new WalletError(`Unsupported payment scheme: ${accepted.scheme}. SmartAccountWallet only supports "exact".`)
    }

    try {
      if (serverAuthorization) {
        return await createCoSignedPayment(instruction, this.config, serverAuthorization)
      }
      return await createExactPayment(instruction, this.config)
    } catch (error) {
      throw new WalletError(
        `Failed to create smart account payment: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined,
      )
    }
  }

  get address(): Address {
    return this.config.smartAccountAddress
  }
}

File: rust/ampersend-sdk/src/x402/wallets/smart_account/wallet.rs
use alloy_primitives::Address;
use async_trait::async_trait;

use super::exact::create_exact_payment;
use crate::error::WalletError;
use crate::smart_account::constants::OWNABLE_VALIDATOR;
use crate::x402::types::{PaymentPayload, PaymentRequirements};
use crate::x402::wallet::X402Wallet;

/// Configuration for SmartAccountWallet.
#[derive(Debug, Clone)]
pub struct SmartAccountConfig {
    /// Smart account address
    pub smart_account_address: Address,
    /// Session key private key for signing (hex-encoded with 0x prefix)
    pub session_key_private_key: String,
    /// Chain ID for the blockchain network
    pub chain_id: u64,
    /// OwnableValidator address (defaults to standard OwnableValidator)
    pub validator_address: Option<Address>,
}

/// SmartAccountWallet — Smart account wallet implementation using ERC-1271.
///
/// Creates payment payloads signed by a smart account using ERC-1271 standard.
/// Supports Safe accounts with OwnableValidator module.
/// Only supports the "exact" payment scheme with ERC-3009 (USDC) authorizations.
pub struct SmartAccountWallet {
    config: SmartAccountConfig,
    validator_address: Address,
}

impl SmartAccountWallet {
    pub fn new(config: SmartAccountConfig) -> Self {
        let validator_address = config.validator_address.unwrap_or_else(|| {
            OWNABLE_VALIDATOR
                .parse::<Address>()
                .expect("Invalid OWNABLE_VALIDATOR constant")
        });
        Self {
            config,
            validator_address,
        }
    }

    /// Returns the smart account address.
    pub fn address(&self) -> Address {
        self.config.smart_account_address
    }
}

#[async_trait]
impl X402Wallet for SmartAccountWallet {
    async fn create_payment(
        &self,
        requirements: &PaymentRequirements,
    ) -> Result<PaymentPayload, WalletError> {
        if requirements.scheme != "exact" {
            return Err(WalletError::UnsupportedScheme(requirements.scheme.clone()));
        }

        create_exact_payment(
            requirements,
            self.config.smart_account_address,
            &self.config.session_key_private_key,
            self.config.chain_id,
            self.validator_address,
        )
        .await
    }
}
